File: s5/src/FcAir/fcairstats.cpp
#include "fcairstats.h"

 // Name, command, detUntil, scale, Unit
 const StatsEntry FcAirStats::s_StatsEntries[NumShortStats] =
{
    { "CalTime","Time", "Other", 1, "", StatsEntry::DecodeCalendar  },
    { "UpTime","Uptime", "Other", 1, "", StatsEntry::DecodeUptime  },

    { "FuelCellState", "FuelCellState", "Other", 1, "", StatsEntry::DecodeNormal  },
    { "Load_I", "LoadCurrent", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "HybrdBusV", "HybridBusVoltage", "Power", 100, "V", StatsEntry::DecodeNormal  },
    { "Converter_I", "ConverterCurrent", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "DC1_I", "DC1Current", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "DC2_I", "DC2Current", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "DC3_I", "DC3Current", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "CpuTemp", "CpuTemperature", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Cnvrtr_T1", "ConverterTemperature1", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Cnvrtr_T2", "ConverterTemperature2", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Cnvrtr_T3", "ConverterTemperature3", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Bat_V", "BatteryVoltage", "Power", 100, "V", StatsEntry::DecodeNormal  },
    { "Bat_In_I", "BatteryCurrent", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "Bat_Out_I", "BatteryOutCurrent", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "Supply18V", "Supply18Voltage", "Other", 100, "V", StatsEntry::DecodeNormal  },
    { "Supply18I", "Supply18Current", "Other", 100, "A", StatsEntry::DecodeNormal  },
    { "Amb_T", "Ambient", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "ErrrMsk", "MasterErrorMask", "Other", 1, "", StatsEntry::DecodeHex  },
    { "WrningMsk","MasterWarningMask", "Other", 1, "", StatsEntry::DecodeHex  },

    { "Stk1_V", "fc1Vsns", "Power", 100, "V", StatsEntry::DecodeNormal  },
    { "Stk1_I", "fc1Isns", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "Stk1_H2_Psi","Stk1-FuelPressure", "Other", 100, "PSI", StatsEntry::DecodeNormal },
    { "Stk1BallastPsi", "Stk1-BallastPressure", "Other", 100, "PSI", StatsEntry::DecodeNormal },
    { "Stk1CoolantOut", "Stk1-CoolantOut", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Stk1SurfaceTmp", "Stk1-StackSurface", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Stk1FanDC", "Stk1-AirBlowerDC", "DutyCycle", 100, "%", StatsEntry::DecodeNormal  },
    { "Stk1DiverterDC", "Stk1-DiverterDC", "DutyCycle", 100, "%", StatsEntry::DecodeNormal  },
    { "Stk1StateMachine", "Stk1-StateMachine", "Other", 1, "", StatsEntry::DecodeNormal  },
    { "Stk1ErrMsk", "Stk1-ErrorMask", "Other", 1, "", StatsEntry::DecodeHex  },
    { "Stk1WrngMsk", "Stk1-WarningMask", "Other", 1, "", StatsEntry::DecodeHex  },

    { "Stk2-V", "fc2Vsns", "Power", 100, "V", StatsEntry::DecodeNormal  },
    { "Stk2-I", "fc2Isns", "Power", 100, "A", StatsEntry::DecodeNormal  },
    { "Stk2H2Psi", "Stk2-FuelPressure", "Other", 100, "PSI", StatsEntry::DecodeNormal },
    { "Stk2BallastPsi", "Stk2-BallastPressure", "Other", 100, "PSI", StatsEntry::DecodeNormal },
    { "Stk2CoolantOut", "Stk2-CoolantOut", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Stk2SurfaceTmp", "Stk2-StackSurface", "Temperature", 10, "C", StatsEntry::DecodeNormal  },
    { "Stk2FanDC", "Stk2-AirBlowerDC", "DutyCycle", 100, "%", StatsEntry::DecodeNormal  },
    { "Stk2DiverterDC", "Stk2-DiverterDC", "DutyCycle", 100, "%", StatsEntry::DecodeNormal  },
    { "Stk2StateMachine", "Stk2-StateMachine", "Other", 1, "", StatsEntry::DecodeNormal  },
    { "Stk2ErrMsk", "Stk2-ErrorMask", "Other", 1, "", StatsEntry::DecodeHex  },
    { "Stk2WrngMsk", "Stk2-WarningMask", "Other", 1, "", StatsEntry::DecodeHex  },
};

FcAirStats::FcAirStats()
 : BaseStats(StatsType::FcAir)
{
    for(int i = 0; i <NumShortStats; i++){
        //mItems.append(new StatsItem(StatsType::FcAir,i)) ;
        StatsItem *t = new StatsItem(StatsType::FcAir,i);

        #ifdef Q_OS_WIN || Q_OS_LINUX
        t->m_nValue = 11;
#else
        t->m_nValue = 0;
#endif
        mItems.append(t);
    }
}

void FcAirStats::deserialize(const QString &Data){//make this to const ref
// get tick count or time count when data comes in
    char sDelim = { '*' };
     QStringList sNoCs = Data.split(sDelim);

     char sDelim2 = { ',' };
     QStringList sArgs = sNoCs[0].split(sDelim2);
     int nSkip = 5;
     double nVal;
     int nStatsSkip = 2;

     for (int i = nSkip; i < sArgs.size(); i++)
     {
         nVal = 0;
         //try
         //{
             if (sArgs[i] != "")
                 nVal = sArgs[i].toDouble();
         //}
        // catch { }

         mItems[i - nSkip + nStatsSkip]->m_nValue = (int)(nVal * (double)s_StatsEntries[i - nSkip + nStatsSkip].nScale);
         if (i - nSkip + nStatsSkip + 1 >= NumShortStats)
             break;
     }
     //emit conversionComplete();
}

QString FcAirStats::getFavName(int idx) const
{
  if(m_pfavList.size() > idx)
    return s_StatsEntries[m_pfavList.at(idx)->m_npos].name; // what if we have fcstats and rwsstats come in ?
  else
      return "items ERROR";
}

QString FcAirStats::getFavUnit(int idx) const
{
    if(m_pfavList.size() > idx)
        return s_StatsEntries[m_pfavList.at(idx)->m_npos].unit;  // what if we have fcstats and rwsstats come in ?
    else
        return "Index ERROR";
}

QString FcAirStats::getName(int idx) const{
      if((mItems.size() > idx) && (idx >= 0)){
          return s_StatsEntries[idx].name;
      }
    return QString("%");
}
QString FcAirStats::getUnit(int idx) const{
    if((mItems.size() > idx) && (idx >= 0)){
        return s_StatsEntries[idx].unit;
    }
    return QString(" ");
}

void FcAirStats::setFav(bool val, int idx)
{
    if(mItems.size() > idx){
      //if(mItems.at(idx)->m_bFav != val){
        mItems.at(idx)->m_bFav = val;
       // emit favChanged(this, idx);
       /*
        if(mItems.at(idx)->m_bFav)
          emit favAdded(this,idx);
        else
          emit favRemoved(this,idx);
          */
      //}
    }
}

void FcAirStats::setForce(bool val, int idx)
{// should we expect for a reply ?
    if(mItems.size() > idx){
        mItems.at(idx)->m_bForce = val;
        if(mItems.at(idx)->m_bForce)
           emit sendForceCommand("force " + s_StatsEntries[idx].command);
        else
           emit sendForceCommand("clear " + s_StatsEntries[idx].command);
    }
}

FcAirStats* FcAirStats::getFcStatsInstance() {
    return this;
}

File: s5/src/FcAir/FcAirStats.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace UpgenConsole.FcAir
{
    public class FcAirStats : BaseStats
    {
	    public int [] m_IntGroup;
        public int [] m_ShortGroup;
        public int m_TickCount;

        public enum IntStats {
            CalendarTime,
            SystemUptime,

            NumIntStats,
        };

        static public String [] m_Units =
        {
            "secs",  //CalendarTime
            "secs",  //UpTime

            "",  //FuelCellState,
            "A",  //LoadCurrent,
            "V",  //HybridBusVoltage,
            "A",  //ConverterCurrent,
            "A",  //DC1Current,
            "A",  //DC2Current,
            "A",  //DC3Current,
            "C",  //CpuTemperature,
            "C",  //ConverterTemperature1,
            "C",  //ConverterTemperature2,
            "C",  //ConverterTemperature3,
            "V",  //BatteryVoltage,
            "A",  //BatteryInCurrent,
            "A",  //BatteryOutCurrent,
            "V",  //Supply18Voltage,
            "A",  //Supply18Current,
            "C",  //AmbientTemperature,
            "",  //StatusErrorMask,
            "",  //StatusWarningMask,		
		    "psi",//TankPsi,
            "g",//H2MassRemaining,
            "min",//TimeRemaining,

            "V",  //Stk1FuelCellVoltage,	// 100ths
            "A",  //Stk1FuelCellCurrent,
            "psi",  //Stk1HydrogenPressure,
            "psi",  //Stk1BallastPressure,
            "C",  //Stk1CoolantOutputTemp,
            "C",  //Stk1StackSurfaceTemp,
            "%",  //Stk1AirBlowerDriveSignal,
            "%",  //Stk1BypassDriveSignal,	
            "",  //Stk1StateMachineStatus,
            "",  //Stk1ErrorMask,
            "",  //Stk1WarningMask,
		
            "V",  //Stk2FuelCellVoltage,	// 100ths
            "A",  //Stk2FuelCellCurrent,
            "psi",  //Stk2HydrogenPressure,
            "psi",  //Stk2BallastPressure,
            "C",  //Stk2CoolantOutputTemp,
            "C",  //Stk2StackSurfaceTemp,
            "%",  //Stk2AirBlowerDriveSignal,
            "%",  //Stk2BypassDriveSignal,	
            "",  //Stk2StateMachineStatus,
            "",  //Stk2ErrorMask,
            "",  //Stk2WarningMask,
        };

        static public int[] m_Scale =
        {
            1,  //CalendarTime
            1,  //UpTime
            //                              ICD     ADCTask         Old ADC
            1,  //FuelCellState,
            100,  //LoadCurrent,            100     10              1
            100,  //HybridBusVoltage,       100     10              1
            100,  //ConverterCurrent,       100     Not Correct
            100,  //DC1Current,             100     Not Correct
            100,  //DC2Current,             100     Not Correct
            100,  //DC3Current,             100     Not Correct
            10,  //CpuTemperature,          10      1
            10,  //ConverterTemperature1,    10      Not Correct
            10,  //ConverterTemperature2,    10      Not Correct
            10,  //ConverterTemperature3,    10      Not Correct
            100,  //BatteryVoltage,         100     10              1
            100,  //BatteryInCurrent,       100     10
            100,  //BatteryOutCurrent,      100     10
            100,  //Supply18Voltage,         100     10              1
            100,  //Supply18Current,        100     10
            10,  //AmbientTemperature,      10      1
            1,  //StatusErrorMask,
            1,  //StatusWarningMask,
		    10, //TankPsi,
            1,  //H2MassRemaining,
            1,   //TimeRemaining,
		
            100,  //Stk1FuelCellVoltage,	100     10
            100,  //Stk1FuelCellCurrent,    100     10
            100,  //Stk1HydrogenPressure,   100     10              1
            100,  //Stk1BallastPressure,    100     10              1
            10,  //Stk1CoolantOutputTemp,   10      1
            10,  //Stk1StackSurfaceTemp,    10      1
            1,  //Stk1AirBlowerDriveSignal, 100     N/A
            1,  //Stk1BypassDriveSignal,	100     N/A
            1,  //Stk1StateMachineStatus,   
            1,  //Stk1ErrorMask,
            1,  //Stk1WarningMask,
		
            100,  //Stk2FuelCellVoltage,    100     10
            100,  //Stk2FuelCellCurrent,    100     10
            100,  //Stk2HydrogenPressure,   100     10              1
            100,  //Stk2BallastPressure,    100     10              1
            10,  //Stk2CoolantOutputTemp,   10      1
            10,  //Stk2StackSurfaceTemp,    10      1
            1,  //Stk2AirBlowerDriveSignal, 100     N/A
            1,  //Stk2BypassDriveSignal,	100     N/A
            1,  //Stk2StateMachineStatus,
            1,  //Stk2ErrorMask,
            1,  //Stk2WarningMask,
        };

        public enum ShortStats
        {
            CalendarTime,
            UpTime,

            FuelCellState,
            LoadCurrent,
            HybridBusVoltage,
            ConverterCurrent,
            DC1Current,
            DC2Current,
            DC3Current,
            CpuTemperature,
            ConverterTemperature1,
            ConverterTemperature2,
            ConverterTemperature3,
            BatteryVoltage,
            BatteryInCurrent,
            BatteryOutCurrent,
            Supply18Voltage,
            Supply18Current,
            AmbientTemperature,
            StatusErrorMask,
            StatusWarningMask,
		    TankPsi,
            H2MassRemaining,
            TimeRemaining,

            Stk1FuelCellVoltage,	// 100ths
            Stk1FuelCellCurrent,
            Stk1HydrogenPressure,
            Stk1BallastPressure,
            Stk1CoolantOutputTemp,
            Stk1StackSurfaceTemp,
            Stk1AirBlowerDriveSignal,
            Stk1BypassDriveSignal,
            Stk1StateMachineStatus,
            Stk1ErrorMask,
            Stk1WarningMask,

            Stk2FuelCellVoltage,	// 100ths
            Stk2FuelCellCurrent,
            Stk2HydrogenPressure,
            Stk2BallastPressure,
            Stk2CoolantOutputTemp,
            Stk2StackSurfaceTemp,
            Stk2AirBlowerDriveSignal,
            Stk2BypassDriveSignal,
            Stk2StateMachineStatus,
            Stk2ErrorMask,
            Stk2WarningMask,

            NumShortStats,
        };

        public FcAirStats()
            : base(StatsType.FcAir)
        {
            m_IntGroup = new Int32[(int)IntStats.NumIntStats];
            m_ShortGroup = new Int32[(int)ShortStats.NumShortStats];
        }

        public static FcAirStats deserialize(String Result)
        {
            FcAirStats stats = new FcAirStats();

            stats.m_TickCount = System.Environment.TickCount;

            char[] sDelim = { '*' };
            string[] sNoCs = Result.Split(sDelim);

            char[] sDelim2 = { ',' };
            string[] sArgs = sNoCs[0].Split(sDelim2);
            int nSkip = 5;
            double nVal;
            int nStatsSkip = 2;

            for (int i = nSkip; i < sArgs.Length; i++)
            {
                nVal = 0;
                try
                {
                    if (sArgs[i] != "")
                        nVal = Convert.ToDouble(sArgs[i]);

                }
                catch { }

                stats.m_ShortGroup[i - nSkip + nStatsSkip] = Convert.ToInt32(nVal * (double)m_Scale[i - nSkip + nStatsSkip]);
                if (i - nSkip + nStatsSkip + 1 >= stats.getNumStats())
                    break;
            }


            return stats;
        }

        public static FcAirStats deserialize(byte[] pPkt)
        {
            FcAirStats stats = new FcAirStats();

            stats.m_TickCount = System.Environment.TickCount;

            int nOffset = 2;
            for (int i = 0; i < 2; i++)
            {
                int nVal = SerialBinaryCommand.getIntFromStream(pPkt, ref nOffset);
                stats.m_ShortGroup[i] = nVal;
            }
            for (int i = 2; i < stats.m_ShortGroup.Length; i++)
            {
                short shVal = SerialBinaryCommand.getShortFromStream(pPkt, ref nOffset);
                stats.m_ShortGroup[i] = (int)shVal;
            }

            return stats;
        }

        private int getShortValue(int idx)
        {
            return m_ShortGroup[idx];
        }

        private int getShortScale(int idx)
        {
            return m_Scale[idx];
        }

        public override double getDoubleValue(int item)
        {
            double val = (double)getShortValue(item);
            val /= (double)getShortScale(item);
            return val;
        }

        public override String getStatLabel(int idx)
        {
            return m_Type.ToString() + "." + ((ShortStats)idx).ToString();
        }
        public override String getUnitsLabel(int idx)
        {
            return m_Units[idx];
        }

        public override int getNumStats()
        {
            return (int)m_ShortGroup.Length;
        }

        public override int getTickCount()
        {
            return m_TickCount;
        }

        public override int getSystemUptime()
        {
            return m_IntGroup[(int)IntStats.SystemUptime];
        }
    }
}
